### PR TITLE
Core: Migrating from joda to java.time. Monitoring plugin

### PR DESCRIPTION
**Pull request by pgomulka**
monitoring plugin migration from joda to java.time

refers #27330

### CODE REVIEW

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java [hunk 1]
```diff
@@ -258,9 +257,9 @@ private void checkMonitoringDocs() {
         if (customTimeFormat == null) {
             customTimeFormat = "YYYY.MM.dd";
         }
-
-        DateTimeFormatter dateParser = ISODateTimeFormat.dateTime().withZoneUTC();
-        DateTimeFormatter dateFormatter = DateTimeFormat.forPattern(customTimeFormat).withZoneUTC();
+//TODO confirm this is the right format
+        DateFormatter dateParser = DateFormatter.forPattern("strict_date_time");
```
Comment (pgomulka) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java]: maybe it should be strict_date_optional_time_nanos ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/monitoring/exporter/MonitoringDoc.java [hunk 2]
```diff
@@ -123,7 +123,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      * @return a string representing the timestamp
      */
     public static String toUTC(final long timestamp) {
-        return new DateTime(timestamp, DateTimeZone.UTC).toString();
```
Comment (pgomulka) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/monitoring/exporter/MonitoringDoc.java]: joda toString:
     ` Output the date time in ISO8601 format (yyyy-MM-ddTHH:mm:ss.SSSZZ).`
when java-time
     ` {@code 2007-12-03T10:15:30+01:00[Europe/Paris]}.`
but it also says:      ` The output is compatible with ISO-8601 if the offset and ID are the same.`
The quick test for sample long shows that the format is the same
`Instant.ofEpochMilli(123).atZone(ZoneOffset.UTC).toString()`
`new DateTime(123, DateTimeZone.UTC).toString();`
`1970-01-01T00:00:00.123Z`

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java [hunk 3]
```diff
@@ -258,9 +257,9 @@ private void checkMonitoringDocs() {
         if (customTimeFormat == null) {
             customTimeFormat = "YYYY.MM.dd";
         }
-
-        DateTimeFormatter dateParser = ISODateTimeFormat.dateTime().withZoneUTC();
-        DateTimeFormatter dateFormatter = DateTimeFormat.forPattern(customTimeFormat).withZoneUTC();
+//TODO confirm this is the right format
```
Comment (pgomulka) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java]: looks ok to me..
ISODateTimeFormat:
yyyy-MM-dd'T'HH:mm:ss.SSSZZ
when 
DateFormatters.STRICT_DATE_TIME
 (yyyy-MM-dd'T'HH:mm:ss.SSSZZ).

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java [hunk 4]
```diff
@@ -259,8 +258,8 @@ private void checkMonitoringDocs() {
             customTimeFormat = "YYYY.MM.dd";
         }
 
-        DateTimeFormatter dateParser = ISODateTimeFormat.dateTime().withZoneUTC();
-        DateTimeFormatter dateFormatter = DateTimeFormat.forPattern(customTimeFormat).withZoneUTC();
+        DateFormatter dateParser = DateFormatter.forPattern("strict_date_time");
```
Comment (pgomulka) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporterIntegTests.java]: or maybe `DateTimeFormatter.ISO_DATE_TIME` ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java [hunk 5]
```diff
@@ -31,7 +31,7 @@ public Authentication authentication() {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        if (out.getVersion().before(Version.V_6_6_0)) {
+        if (out.getVersion().before(Version.V_7_0_0)) {
```
Comment (danielmitterdorfer) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java]: Why is this changed in this PR? This seems unrelated to me.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java [hunk 6]
```diff
@@ -41,7 +41,7 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
-        if (in.getVersion().before(Version.V_6_6_0)) {
+        if (in.getVersion().before(Version.V_7_0_0)) {
```
Comment (danielmitterdorfer) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java]: Why is this changed in this PR? This seems unrelated to me.

**Review by danielmitterdorfer — COMMENTED** [review]
Looks fine overall but I think I spotted one unrelate change.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java (hunk 5 above)
Comment (pgomulka) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java]: that must have been some debugging remains
removing

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java (hunk 6 above)
Comment (pgomulka) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/AuthenticateResponse.java]: removing

**Review by danielmitterdorfer — APPROVED** [review]
LGTM.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/monitoring/exporter/MonitoringDoc.java (hunk 2 above)
Comment (spinscale) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/monitoring/exporter/MonitoringDoc.java]: can we make this more explicit and not rely on the `toString()` method of classes we dont have control over (even though I assume this does not get changed) Using a `strict_date_time` formatter might just work as expected?

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java [hunk 7]
```diff
@@ -57,7 +57,9 @@ public CleanerService(Settings settings, ClusterSettings clusterSettings, Thread
     @Override
     protected void doStart() {
         logger.debug("starting cleaning service");
-        threadPool.schedule(executionScheduler.nextExecutionDelay(new DateTime(ISOChronology.getInstance())), executorName(), runnable);
+        threadPool.schedule(executionScheduler.nextExecutionDelay(ZonedDateTime.now(Clock.systemDefaultZone())),
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java]: can we use `Clock.systemUTC()` here? I see that we did not do this properly before though... see the DateTime() ctor

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java [hunk 8]
```diff
@@ -191,7 +193,7 @@ protected void doRunInLifecycle() throws Exception {
          */
         @Override
         protected void onAfterInLifecycle() {
-            DateTime start = new DateTime(ISOChronology.getInstance());
+            ZonedDateTime start = ZonedDateTime.now(Clock.systemDefaultZone());
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java]: `Clock.systemUTC()`?

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java [hunk 9]
```diff
@@ -196,48 +195,48 @@ protected void createWatcherHistoryIndex(final DateTime creationDate) {
     /**
      * Creates a watcher history index from the specified version.
      */
-    protected void createWatcherHistoryIndex(final DateTime creationDate, final String version) {
-        final DateTimeFormatter formatter = DateTimeFormat.forPattern("YYYY.MM.dd").withZoneUTC();
-        final String index = ".watcher-history-" + version + "-" + formatter.print(creationDate.getMillis());
+    protected void createWatcherHistoryIndex(final ZonedDateTime creationDate, final String version) {
+        DateFormatter formatter = DateFormatter.forPattern("YYYY.MM.dd").withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: `YYYY` in java time means week based year, where is in joda time it means year of era - this is lower case `yyyy` in java time

see https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
see https://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeFormat.html

Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: does it make sense to make the formatter static?

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java [hunk 10]
```diff
@@ -196,48 +195,48 @@ protected void createWatcherHistoryIndex(final DateTime creationDate) {
...
     }
 
     /**
      * Creates a monitoring timestamped index using a given template version.
      */
-    protected void createTimestampedIndex(DateTime creationDate, String version) {
-        final DateTimeFormatter formatter = DateTimeFormat.forPattern("YYYY.MM.dd").withZoneUTC();
-        final String index = ".monitoring-es-" + version + "-" + formatter.print(creationDate.getMillis());
+    protected void createTimestampedIndex(ZonedDateTime creationDate, String version) {
+        DateFormatter formatter = DateFormatter.forPattern("YYYY.MM.dd").withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: same comments as above, static and week based year?

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerServiceTests.java [hunk 11]
```diff
@@ -129,20 +131,23 @@ public void testSetGlobalRetentionAppliesEvenIfLicenseDisallows() {
...
+        now = ZonedDateTime.of(2015, 1, 1, 0, 59, 0, 0, ZoneOffset.UTC);
         assertThat(scheduler.nextExecutionDelay(now).millis(), equalTo(TimeValue.timeValueMinutes(1).millis()));
 
-        now = new DateTime(2015, 1, 1, 23, 59, DateTimeZone.UTC);
+        now = ZonedDateTime.of(2015, 1, 1, 23, 59, 0, 0, ZoneOffset.UTC);
         assertThat(scheduler.nextExecutionDelay(now).millis(), equalTo(TimeValue.timeValueMinutes(60 + 1).millis()));
 
-        now = new DateTime(2015, 1, 1, 12, 34, 56);
-        assertThat(scheduler.nextExecutionDelay(now).millis(), equalTo(new DateTime(2015, 1, 2, 1, 0, 0).getMillis() - now.getMillis()));
+        ZoneId defaultZone = Clock.systemDefaultZone().getZone();
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerServiceTests.java]: if above is changed to UTC this will need to change as well...

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java (hunk 9 above)
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: does it make sense to add a test for this corner case in order to catch it?

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringTemplateUtilsTests.java [hunk 12]
```diff
@@ -93,9 +92,10 @@ public void testEmptyPipeline() throws IOException {
     }
 
     public void testIndexName() {
-        final long timestamp = new DateTime(2017, 8, 3, 13, 47, 58, DateTimeZone.UTC).getMillis();
+        final long timestamp = ZonedDateTime.of(2017, 8, 3, 13, 47, 58,
+            0, ZoneOffset.UTC).toInstant().toEpochMilli();
 
-        DateTimeFormatter formatter = DateTimeFormat.forPattern("YYYY.MM.dd").withZoneUTC();
+        DateFormatter formatter = DateFormatter.forPattern("YYYY.MM.dd").withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringTemplateUtilsTests.java]: week based formatter

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringTemplateUtilsTests.java [hunk 13]
```diff
@@ -105,7 +105,7 @@ public void testIndexName() {
         assertThat(indexName(formatter, MonitoredSystem.BEATS, timestamp),
                 equalTo(".monitoring-beats-" + TEMPLATE_VERSION + "-2017.08.03"));
 
-        formatter = DateTimeFormat.forPattern("YYYY-dd-MM-HH.mm.ss").withZoneUTC();
+        formatter = DateFormatter.forPattern("YYYY-dd-MM-HH.mm.ss").withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/MonitoringTemplateUtilsTests.java]: week based formatter

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporterIT.java [hunk 14]
```diff
@@ -352,7 +354,8 @@ public void testDynamicIndexFormatChange() throws Exception {
                                remoteClusterAllowsWatcher, currentLicenseAllowsWatcher, watcherAlreadyExists);
         MockRequest recordedRequest = assertBulk(webServer);
 
-        String indexName = indexName(DateTimeFormat.forPattern("YYYY.MM.dd").withZoneUTC(), doc.getSystem(), doc.getTimestamp());
+        DateFormatter formatter = DateFormatter.forPattern("YYYY.MM.dd").withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporterIT.java]: week based formatter

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/integration/MonitoringIT.java [hunk 15]
```diff
@@ -277,9 +278,10 @@ private void assertMonitoringDoc(final Map<String, Object> document,
 
         assertThat(((Number) source.get("interval_ms")).longValue(), equalTo(interval.getMillis()));
 
-        assertThat(index, equalTo(MonitoringTemplateUtils.indexName(DateTimeFormat.forPattern("YYYY.MM.dd").withZoneUTC(),
-                                                                    expectedSystem,
-                                                                    ISODateTimeFormat.dateTime().parseMillis(timestamp))));
+        DateFormatter formatter = DateFormatter.forPattern("YYYY.MM.dd");
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/integration/MonitoringIT.java]: week based formatter

**Review by spinscale — CHANGES_REQUESTED** [review]
left a few comments, mostly about week year based format that should be replaced

File: x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java (hunk 9 above)
Comment (pgomulka) [reply x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: hmm this is a test class, what do you suggest  to catch?

Comment (spinscale) [reply x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/cleaner/AbstractIndicesCleanerTestCase.java]: indeed, nevermind.

**Comment by spinscale** [comment]
I think the most important date was not changed in the PR and this can suffer from the weekyear issue: the formatter string of the exporter that sends the data to the local cluster or a remote one.

I put this test in `org.elasticsearch.xpack.monitoring.exporter.ExportersTests` and it fails

```
    public void testDateFormatter() {
        Exporter.Config config = mock(Exporter.Config.class);
        when(config.name()).thenReturn("anything");
        when(config.settings()).thenReturn(Settings.EMPTY);
        DateFormatter formatter = Exporter.dateTimeFormatter(config);
        Instant instant = Instant.ofEpochSecond(randomLongBetween(0, 86400 * 365 * 130L));
        ZonedDateTime zonedDateTime = instant.atZone(ZoneOffset.UTC);
        int year = zonedDateTime.getYear();
        int month = zonedDateTime.getMonthValue();
        int day = zonedDateTime.getDayOfMonth();
        String expecdateDate = String.format(Locale.ROOT, "%02d.%02d.%02d", year, month, day);
        String formattedDate = formatter.format(instant);
        assertThat("input date was " + instant, expecdateDate, is(formattedDate));
    }
```

I think we need to fix `Exporter.INDEX_FORMAT`

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java [hunk 16]
```diff
@@ -242,14 +243,16 @@ public void cancel() {
...
-            DateTime next = now.withTimeAtStartOfDay().plusHours(1);
+            ZonedDateTime next = now.toLocalDate()
+                .atStartOfDay(now.getZone())
+                .plusHours(1);
             // if it's not after now, then it needs to be the next day!
             if (next.isAfter(now) == false) {
                 next = next.plusDays(1);
             }
-            return TimeValue.timeValueMillis(next.getMillis() - now.getMillis());
+            return TimeValue.timeValueMillis(next.toInstant().toEpochMilli() - now.toInstant().toEpochMilli());
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/cleaner/CleanerService.java]: minor nit: you could simplify this to `Duration.between(now, next).toMillis()` (pay attention that you have to switch the order in that case)

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporter.java [hunk 17]
```diff
@@ -113,11 +113,11 @@ public void close() {
 
     protected abstract void doClose();
 
-    protected static DateTimeFormatter dateTimeFormatter(final Config config) {
+    protected static DateFormatter dateTimeFormatter(final Config config) {
         Setting<String> setting = INDEX_NAME_TIME_FORMAT_SETTING.getConcreteSettingForNamespace(config.name);
         String format = setting.exists(config.settings()) ? setting.get(config.settings()) : INDEX_FORMAT;
         try {
-            return DateTimeFormat.forPattern(format).withZoneUTC();
+            return DateFormatter.forPattern(format).withZone(ZoneOffset.UTC);
```
Comment (spinscale) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporter.java]: this also requires a change in 6.7, otherwise users using outdated formats will not be warned if they do so, as this uses the joda time API directly

**Review by spinscale — APPROVED** [review]

**Comment by spinscale** [comment]
FYI: I created #38309 for the required 6.7 change
